### PR TITLE
Close ServerTransport instead of the raw connection

### LINKED ISSUE #528: goroutine leaks
stream.NewClientStream seems to leak a goroutine.

I have some unsent local changes to the end2end tests which check for goroutine leaks on each test.

I see:

```
--- FAIL: TestExceedMaxStreamsLimit (7.02s)
        end2end_test.go:365: Running test in TCP environment...
        end2end_test.go:1388: Leaked goroutine: goroutine 2738 [select]:
                google.golang.org/grpc.NewClientStream.func1(0x7f9b8946d248, 0xc82035e690, 0xc8200da2a0, 0xc820013400)
                        /home/bradfitz/src/google.golang.org/grpc/stream.go:148 +0x257
                created by google.golang.org/grpc.NewClientStream
                        /home/bradfitz/src/google.golang.org/grpc/stream.go:156 +0xa99
--- FAIL: TestCompressOK (6.06s)
        end2end_test.go:365: Running test in TCP environment...
        end2end_test.go:1388: Leaked goroutine: goroutine 2750 [select]:
                google.golang.org/grpc.NewClientStream.func1(0x7f9b8946d248, 0xc82035e870, 0xc82010c9a0, 0xc820013860)
                        /home/bradfitz/src/google.golang.org/grpc/stream.go:148 +0x257
                created by google.golang.org/grpc.NewClientStream
                        /home/bradfitz/src/google.golang.org/grpc/stream.go:156 +0xa99
FAIL
FAIL    google.golang.org/grpc/test     33.222s
```

That leak is from `stream.go`:

```go
// NewClientStream creates a new Stream for the client side. This is called
// by generated code.
func NewClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, method string, opts ...CallOption) (ClientStream, error) {
....
....
....
        // Listen on ctx.Done() to detect cancellation when there is no pending
        // I/O operations on this stream.
        go func() {
                select {
                case <-t.Error():
                        // Incur transport error, simply exit.
                case <-s.Context().Done():
                        err := s.Context().Err()
                        cs.finish(err)
                        cs.closeTransportStream(transport.ContextErr(err))
                }
        }()
        return cs, nil
}
```

I started looking for this when running the race detector with a high `go test -race -count=NNN` value, because the race detector hit its trackable goroutine limit and crashed, which made me suspicious what was leaking.

I'll send changes for adding the goroutine leak checker.

### PR DESCRIPTION
**Pull request by iamqizhao**
introduced by #508 

fixes #528

### CODE REVIEW

File: test/end2end_test.go [hunk 1]
```diff
@@ -1354,7 +1354,7 @@ func interestingGoroutines() (gs []string) {
 	return
 }
 
-var failOnLeaks = flag.Bool("fail_on_leaks", false, "Fail tests if goroutines leak.")
+var failOnLeaks = flag.Bool("fail_on_leaks", true, "Fail tests if goroutines leak.")
```
Comment (tamird) [inline test/end2end_test.go]: Maybe just remove this flag entirely? I can't imagine when you'd want to run without it.

**Comment by tamird** [comment]
LGTM

File: test/end2end_test.go (hunk 1 above)
Comment (iamqizhao) [inline test/end2end_test.go]: done.

**Comment by tamird** [comment]
:ship:

**Comment by bradfitz** [comment]
Whoops, sorry!

LGTM

**Comment by tamird** [comment]
@iamqizhao you should probably cancel this build: https://travis-ci.org/grpc/grpc-go/builds/107889107 since it's for the first commit in this PR

**Pull request merged**